Modules/Test/classes/ScoreReporting/class.ilObjTestScoreSettings.php

```diff
@@ -0,0 +1,851 @@
+    {
+        $clone = clone $this;
+        $clone->highscore_top_table = $highscore_top_table;
+        return $clone;
+    }
+
+    public function getHighScoreMode(): int
+    {
+        if ($this->getHighscoreTopTable() && $this->getHighscoreOwnTable()) {
+            return self::HIGHSCORE_SHOW_ALL_TABLES;
+        }
+
+        if ($this->getHighscoreTopTable()) {
+            return self::HIGHSCORE_SHOW_TOP_TABLE;
+        }
+
+        if ($this->getHighscoreOwnTable()) {
+            return self::HIGHSCORE_SHOW_OWN_TABLE;
+        }
+
+        return 0;
+    }
+
+    public function getHighscoreTopNum(): int
+    {
+        return $this->highscore_top_num;
+    }
+    public function withHighscoreTopNum(int $highscore_top_num): self
+    {
+        $clone = clone $this;
+        $clone->highscore_top_num = $highscore_top_num;
+        return $clone;
+    }
+
+    public function getHighscoreAnon(): bool
+    {
+        return $this->highscore_anon;
+    }
+    public function withHighscoreAnon(bool $highscore_anon): self
+    {
+        $clone = clone $this;
+        $clone->highscore_anon = $highscore_anon;
+        return $clone;
+    }
+
+    public function getHighscoreAchievedTS(): bool
+    {
+        return $this->highscore_achieved_ts;
+    }
+    public function withHighscoreAchievedTS(bool $highscore_achieved_ts): self
+    {
+        $clone = clone $this;
+        $clone->highscore_achieved_ts = $highscore_achieved_ts;
+        return $clone;
+    }
+
+    public function getHighscoreScore(): bool
+    {
+        return $this->highscore_score;
+    }
+    public function withHighscoreScore(bool $highscore_score): self
+    {
+        $clone = clone $this;
+        $clone->highscore_score = $highscore_score;
+        return $clone;
+    }
+
+    public function getHighscorePercentage(): bool
+    {
+        return $this->highscore_percentage;
+    }
+    public function withHighscorePercentage(bool $highscore_percentage): self
+    {
+        $clone = clone $this;
+        $clone->highscore_percentage = $highscore_percentage;
+        return $clone;
+    }
+
+    public function getHighscoreHints(): bool
+    {
+        return $this->highscore_hints;
+    }
+    public function withHighscoreHints(bool $highscore_hints): self
+    {
+        $clone = clone $this;
+        $clone->highscore_hints = $highscore_hints;
+        return $clone;
+    }
+
+    public function getHighscoreWTime(): bool
+    {
+        return $this->highscore_wtime;
+    }
+    public function withHighscoreWTime(bool $highscore_wtime): self
+    {
+        $clone = clone $this;
+        $clone->highscore_wtime = $highscore_wtime;
+        return $clone;
+    }
+
+    public function getForm(
+        bool $anonymity_flag,
+        bool $disabled_flag,
+        string $action,
+        \ilLanguage $lng,
+        FormFactory $form_factory,
+        FieldFactory $f,
+        Refinery $refinery
+    ): Form {
```

Comment:
Please do not return a `Form` here, as this will break composability. Return the complete section, including the transformation that makes it output an `ilObjTestScoreSettings`.

The general idea is: All the objects that are filled via forms should only return `Fields`. These can be composed. The form should then be wrapped around the fields that we are interested in the presentation layer to handle the actual in and output. E.g.: We might be thinking about presenting this in some wizard some day. We will then be able to completely reuse this (if it returns a `Field`) and make the decision which form to use at very late point in time.